In [1]:
import pymysql  
import  pymysql.cursors
import numpy as np
import pandas as pd
connection = pymysql.connect(host='localhost',  
                             user='root',  
                             password='woaiyiwen16',  
                             db='sample',  
                             port=3306,  
                             charset='utf8')
cursor = connection.cursor()


i.	What	are	people	saying	about	me	(somebody)?	

In [2]:
sql = "select avg(favorite_count) from 622158584_tweets"
cursor.execute(sql)
result = cursor.fetchall();
result

((Decimal('0.2063'),),)

These data are scrapped from twitter with twitterid = 622158584. By searching #messi, we got three userid who have at least posted with #messi once. So these three at least have one chance to post similar post.
This user got average 0.2063 favoite per post among 223 posts. People barely know him.

ii.	How	viral	are	my	posts?

In [3]:
sql = "select avg(retweet_count) from 622158584_tweets"
cursor.execute(sql)
result = cursor.fetchall();
result 

((Decimal('404.5471'),),)

This user got an average of 404 retweets per post. This is a big number but nor really big in twitter. So this user seems not to be known by others.

iii.	How	much	influence	to	my	posts have?

In [7]:
sql = "select count(retweet_count) from 622158584_tweets where retweet_count>1000 && favorite_count>1000"
cursor.execute(sql)
result = cursor.fetchall();
result 

((0,),)

Favorite count at 1000 is not a really high level in twitter. This user seems to have no tweet meeting this line. So we have to lower the level

In [9]:
sql = "select max(favorite_count), max(retweet_count)from 622158584_tweets"
cursor.execute(sql)
result = cursor.fetchall();
result 

((28, 41285),)

The user got a max favorite count of 28 and a max retweet count of 41285.

In [10]:
sql = "select count(retweet_count) from 622158584_tweets where retweet_count>10000"
cursor.execute(sql)
result = cursor.fetchall();
result 

((1,),)

Since this user got a max retweet of 40k+, we set the limit to 10k and get the result that this large retweet only happened once. The user may not be very influenced among similar users or even not influential at all.

iv.	What	posts are	like	mine?

In [14]:
import pymysql  
import  pymysql.cursors
import numpy as np
import pandas as pd
connection = pymysql.connect(host='localhost',  
                             user='root',  
                             password='woaiyiwen16',  
                             db='sample',  
                             port=3306,  
                             charset='latin1')
cursor = connection.cursor()
sql = "select text from similarposts"
cursor.execute(sql)
result = cursor.fetchall();
result 

(('RT @ForeignPolicy: African soccer players are being exploited and abandoned by foreign governments and ill-intentioned agents.',),
 ('African soccer players are being exploited and abandoned by foreign governments and ill-intentioned agents.',),
 ("RT @elsaetarubia: Fucking hell. Iniesta leaving Barcelona hurts so much. Not only because he's a classy player &amp; one of the best to ever to?",),
 ("RT @DonHaci: Real OG, carried Messi for years, didn't get the Ballon D'or over him when he actually deserved it, on my list along with Puyo?",),
 ('@Holy123Spirit For sure. I?m a Madrid fan, but Iniesta was probably my favorite Barcelona player to watch excluding Messi',),
 ('RT @deejayfaremi: At the peak of their powers, PLAYING TOGETHER, if you had to bin one of Messi, Iniesta, or Xavi from that Barcelona team,?',),
 ("RT @TheBeardedRauI: Barcelona haven't replaced Xavi yet and Iniesta is leaving. Left with Messi and Busquets who'll leave soon maybe next s?",),
 ('My favorite player at B

Assume a user is using hashtag messi,barcelona. We can search tweets by hashtag [messi,barcelona]. All the results are stored in the file named similarposts. So these tweets are using at least two same tags with the user. These posts are like those of the user.

v.	What	users	post like	me?

In [15]:
sql = "select user_screen_name from similarposts"
cursor.execute(sql)
result = cursor.fetchall();
result 

(('fabio_adorno',),
 ('ForeignPolicy',),
 ('WhalaySiS',),
 ('msbknight1',),
 ('JoeBuck713',),
 ('FChidike',),
 ('dydynyc',),
 ('oliver815',),
 ('ahmedelozaa',),
 ('rinnengann',),
 ('FreeMahrezV2',),
 ('Camaj14Kristian',),
 ('ianbevitt',),
 ('Davidfitztv',),
 ('BGFUT15',),
 ('9Sport1',),
 ('CofekSports',),
 ('AnMoSportPsych',),
 ('patrickrody7',),
 ('nan_vzla',),
 ('sarpongmic',),
 ('TrentTakeoff',),
 ('mariormejiav15',),
 ('KamenEntchev',),
 ('tango2022',),
 ('sarpongmic',),
 ('EmilySleeman_8',),
 ('fantastlsco',),
 ('souljaramanga',),
 ('dendi_cules',),
 ('SadioManea',),
 ('ntusntus',),
 ('szw003',),
 ('BosniskBosanac',),
 ('KingMufasah',),
 ('MYashiv',),
 ('RunTownFC',),
 ('TheFantasyFooti',),
 ('offsidetwice',),
 ('Kita___K',),
 ('LCuccittini',),
 ('singha4',),
 ('FCYahoo',),
 ('Up614566Fred',),
 ('HenryBushnell',),
 ('Jesse_jiga',),
 ('aachym',),
 ('Barcelonadaily_',),
 ('Official_T4O',),
 ('MESSI_is_life_',),
 ('MESSI_is_life_',),
 ('hurricanesaucey',),
 ('jep_paul',),
 ('Jewfro_1

These are the user screen name of the posts above which are like those of the user. So these peopel are using similar tags with the user. We can assume they post like the user.

vi.	Who	should	I	be	following?


From the question above, we got three users named fabio_adorno and ForeignPolicy and whalaySiS. Through twitter api we can check whether they followed each other.

In [1]:
import twitter
CONSUMER_KEY = 'ilrMBG1phrX3iT3BCP6rbB0uS'
CONSUMER_SECRET = 'C2m5PwgEzgdipZ0kMlCWIvZzFtZ3Ip3y0Rbsv9PAo2wYcUBMN8'
OAUTH_TOKEN = '906874478554165250-LN01g0KGYAm2UDHgnXow3WDHxKtRFCq'
OAUTH_TOKEN_SECRET = 'Yvxbmm2OEkedq8sZ3cPHtjwPW3FVtx65JnWMtJ1drpno4'
auth = twitter.oauth.OAuth(OAUTH_TOKEN, OAUTH_TOKEN_SECRET,
                           CONSUMER_KEY, CONSUMER_SECRET)

twitter_api = twitter.Twitter(auth=auth)
print(twitter_api)

In [2]:
source = "fabio_adorno"
target = "foreignpolicy"

#-----------------------------------------------------------------------
# perform the API query
# twitter API docs: https://dev.twitter.com/rest/reference/get/friendships/show
#-----------------------------------------------------------------------
result = twitter_api.friendships.show(source_screen_name = source,
                                  target_screen_name = target)

#-----------------------------------------------------------------------
# extract the relevant properties
#-----------------------------------------------------------------------
following = result["relationship"]["target"]["following"]
follows   = result["relationship"]["target"]["followed_by"]

print ("%s following %s: %s" % (source, target, follows))
print ("%s following %s: %s" % (target, source, following))

fabio_adorno following foreignpolicy: True
foreignpolicy following fabio_adorno: False


In [3]:
source = "fabio_adorno"
target = "whalaySiS"

#-----------------------------------------------------------------------
# perform the API query
# twitter API docs: https://dev.twitter.com/rest/reference/get/friendships/show
#-----------------------------------------------------------------------
result = twitter_api.friendships.show(source_screen_name = source,
                                  target_screen_name = target)

#-----------------------------------------------------------------------
# extract the relevant properties
#-----------------------------------------------------------------------
following = result["relationship"]["target"]["following"]
follows   = result["relationship"]["target"]["followed_by"]

print ("%s following %s: %s" % (source, target, follows))
print ("%s following %s: %s" % (target, source, following))

fabio_adorno following whalaySiS: False
whalaySiS following fabio_adorno: False


In [4]:
source = "whalaySiS"
target = "foreignpolicy"

#-----------------------------------------------------------------------
# perform the API query
# twitter API docs: https://dev.twitter.com/rest/reference/get/friendships/show
#-----------------------------------------------------------------------
result = twitter_api.friendships.show(source_screen_name = source,
                                  target_screen_name = target)

#-----------------------------------------------------------------------
# extract the relevant properties
#-----------------------------------------------------------------------
following = result["relationship"]["target"]["following"]
follows   = result["relationship"]["target"]["followed_by"]

print ("%s following %s: %s" % (source, target, follows))
print ("%s following %s: %s" % (target, source, following))

whalaySiS following foreignpolicy: False
foreignpolicy following whalaySiS: False


In [14]:
import pymysql  
import  pymysql.cursors
import numpy as np
import pandas as pd
connection = pymysql.connect(host='localhost',  
                             user='root',  
                             password='woaiyiwen16',  
                             db='sample',  
                             port=3306,  
                             charset='utf8')
cursor = connection.cursor()
sql = "SELECT tag1, count(*) FROM gai595406250_tweets group by tag1 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 184),
 ('LIVMCI', 2),
 ('HelloSir', 2),
 ('ImpeachBuhariNow', 1),
 ('AFCvATL', 1),
 ('HalaMadrid', 1),
 ('RESPECT', 1),
 ('accccolade', 1),
 ('LazyNigerianYouths', 1),
 ('Juventus', 1),
 ('UCL', 1),
 ('greatlagosviews', 1),
 ('ARSSOU', 1),
 ('APorLa13', 1),
 ("'Legends", 1))

In [16]:
sql = "SELECT tag2, count(*) FROM gai595406250_tweets group by tag2 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 197), ('EndSARSNow', 1), ('HalaMadrid', 1), ('lagos', 1))

In [17]:
sql = "SELECT tag3, count(*) FROM gai595406250_tweets group by tag3 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 197), ('SackIGPNow', 1), ('RealMadrid', 1), ('eko', 1))

In [19]:
sql = "SELECT tag4, count(*) FROM gai595406250_tweets group by tag4 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 198), ('ReformPoliceNGNow', 1), ('RealJuv', 1))

This is the frequency of hashtags used by whalaysIs. We ca do this on another two users.

In [20]:
sql = "SELECT tag1, count(*) FROM gai46791653_tweets group by tag1 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 176),
 ('LulaLivre', 2),
 ('AmanhecerPorMarielleEAnderson', 2),
 ('PanoramaTVCultura', 1),
 ('AquivoTIB', 1),
 ('Russia', 1),
 ('map', 1),
 ('SuperLidas', 1),
 ('FakeNews', 1),
 ('SP', 1),
 ('eqp', 1),
 ('URGENTE', 1),
 ('Siria', 1),
 ('Eleições2018', 1),
 ('FelizMiércoles', 1),
 ("'Justiça", 1),
 ('Governo', 1),
 ('MariellePresente', 1),
 ('racismo', 1),
 ('NotíciasABC', 1),
 ('arquivoelpais', 1))

In [21]:
sql = "SELECT tag2, count(*) FROM gai46791653_tweets group by tag2 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 194), ('maps', 1), ('USA', 1), ('HumanRights', 1), ('juventude', 1))

In [22]:
sql = "SELECT tag3, count(*) FROM gai46791653_tweets group by tag3 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 196), ('France', 1), ('VidasNegras', 1))

In [23]:
sql = "SELECT tag4, count(*) FROM gai46791653_tweets group by tag4 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 197), ('UK', 1))

Seemly this user fabio_adorno may like travelling.

In [24]:
sql = "SELECT tag1, count(*) FROM gai26792275_tweets group by tag1 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 230),
 ('MeToo', 4),
 ("'mustread", 1),
 ('MerkelVisit', 1),
 ('EdgeOfChaos', 1),
 ('redesign', 1),
 ('Syria', 1),
 ('Merkel', 1))

In [25]:
sql = "SELECT tag2, count(*) FROM gai26792275_tweets group by tag2 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 238), ('Iran', 1), ('budget', 1))

In [26]:
sql = "SELECT tag3, count(*) FROM gai26792275_tweets group by tag3 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 238), ('Russia', 1), ('policy', 1))

Looks like user fabio_adorno should follow ForeignPolicy even they are familiar in other domain than la liga. And yes, fabio_adorno has followed ForeignPolicy. He made a right choice. The users here can be replaced by any user and we can find the most similar ones.

vii.	What	topics	are	trending	in	my	domain?

Our domain is set to be La Liga. So we search by la liga and top three la liga player to make sure the posts retrieved are in the domain.

In [ ]:
for tweet in tweepy.Cursor (api.search, q = ['la liga','messi','Messi','Ronaldo','ronaldo'], lang = "en").items(2000): 
    results.append(tweet)
    

We found that our domain itself is not often used as a tag. So we search by the top 2 player in La Liga Messi and Ronaldo. So we search by messi and ronaldo and count the tags. 

In [6]:
import pymysql  
import  pymysql.cursors
import numpy as np
import pandas as pd
connection = pymysql.connect(host='localhost',  
                             user='root',  
                             password='woaiyiwen16',  
                             db='sample',  
                             port=3306,  
                             charset='utf8')
cursor = connection.cursor()
sql = "select * from hashcount1"
cursor.execute(sql)
result = cursor.fetchall()
result

(('Ronaldo', 22),
 ('Messi', 19),
 ('messi', 18),
 ('barca', 10),
 ('ballondor', 9),
 ('LIVROM', 9),
 ('fifa18', 9),
 ('barcelona', 8),
 ('ronaldo', 8),
 ('TISNews', 8),
 ('Salah', 7),
 ('freekicks', 6),
 ('wenger', 6),
 ('chelseafc', 6),
 ('backflip', 6),
 ('footballbible', 6),
 ('freestyler', 6),
 ('MoSalah', 5),
 ('salah', 5),
 ('realmadrid', 4),
 ('FIFAWorldCup', 4),
 ('inesta', 4),
 ('football', 4),
 ('FifaWorldCup2018', 3),
 ('IAMKPressMeet', 3),
 ('AttitudeDikha', 3),
 ('legend', 3),
 ('Cristiano', 3),
 ('Crazy', 3),
 ('ModiInChina', 3),
 ('WorldCup', 3),
 ('iniesta', 3),
 ('UCLfinal', 3),
 ('Videos', 3),
 ('APCyberSecuritySummit', 3),
 ('zidane', 3),
 ('Funny', 3),
 ('ZomatoGoldRush', 3),
 ('FIFA', 3),
 ('MessiReyBarça', 2),
 ('LFC', 2),
 ('Infinit8Iniesta', 2),
 ('BAYRMA', 2),
 ('lfc', 2),
 ('FootTheBall', 2),
 ('Liverpool', 2),
 ('cristiano', 2),
 ('fifa', 2),
 ('GraciasIniesta', 2),
 ('penalty', 2),
 ('besoccerEn', 1),
 ('rooney', 1),
 ('New_tricks', 1),
 ('Fact', 1),
 ('DeB

In [8]:
sql = "select * from hashcountsample"
cursor.execute(sql)
result = cursor.fetchall()
result

(('UPDATE', 32),
 ('FCB', 31),
 ('Barcelona', 25),
 ('barcelona', 20),
 ('Chelsea', 20),
 ('messi', 18),
 ('CFC', 17),
 ('trademark', 13),
 ('Barca', 12),
 ('Ronaldo', 10),
 ('barca', 10),
 ('wenger', 8),
 ('freestyler', 8),
 ('chelseafc', 8),
 ('backflip', 8),
 ('footballbible', 8),
 ('Iniesta', 8),
 ('freekicks', 8),
 ('AndresIniesta', 6),
 ('FNH', 6),
 ('OnThisDay', 6),
 ('fcb', 6),
 ('iniesta', 6),
 ('Cule', 5),
 ('LionelMessi', 5),
 ('LFC', 4),
 ('Argentina', 4),
 ('Xavi', 4),
 ('GraciasIniesta', 4),
 ('GraciasAndrés', 4),
 ('fcbarcelona', 4),
 ('legend', 3),
 ('realmadrid', 3),
 ('ballondor', 3),
 ('laliga', 3),
 ('Sports', 3),
 ('zidane', 3),
 ('Europe', 3),
 ('inesta', 3),
 ('FUTBUL', 2),
 ('Madrid', 2),
 ('BarcaMadrid', 2),
 ('FCBARCELONA', 2),
 ('BARvsLEG', 2),
 ('La', 2),
 ('EU', 2),
 ('spain', 2),
 ('recentthrowback', 2),
 ('Throwback', 2),
 ('RealMadrid', 2),
 ('IP', 2),
 ('Fc', 2),
 ('FIFA18', 2),
 ('football', 2),
 ('fifa', 2),
 ('PSG', 2),
 ('LaLigaSantander', 2),
 ('Fa

We can see the toppest ones are the trends in the domain. By adding player and team in the search criteria, we can get more tweets and are more likely to fetch waht's really trending.

viii.	What	keywords/	hashtags	should	I	add	to	my	post?

Assume i'm using #messi very often. I can search by messi and count all the hashtags using with messi as a trend. In a period, these keywords/hashtags should i add to my post.

In [2]:
import pymysql  
import  pymysql.cursors
import numpy as np
import pandas as pd
connection = pymysql.connect(host='localhost',  
                             user='root',  
                             password='woaiyiwen16',  
                             db='sample',  
                             port=3306,  
                             charset='latin1')
cursor = connection.cursor()
sql = "select * from countmessi"
cursor.execute(sql);
result = cursor.fetchall()
result

(('Infinit8Iniesta', 24),
 ('LFC', 13),
 ('OnThisDay', 12),
 ('messi', 9),
 ('fifa18', 9),
 ('Iniesta', 9),
 ('Fraud', 7),
 ('WeAreMessi', 7),
 ('Ads', 7),
 ('Barcelona', 7),
 ('Grab', 7),
 ('Mobiuz', 7),
 ('Advertising', 7),
 ('Taxi', 7),
 ('Uber', 7),
 ('Bus', 7),
 ('PlanB', 6),
 ('InternationalWorkersDay', 6),
 ('Xavi', 5),
 ('iniesta', 5),
 ('xavi', 5),
 ('fintech', 5),
 ('GraciasAndrés', 5),
 ('gerrard', 4),
 ('GraciasIniesta', 4),
 ('UCL', 4),
 ('innovation', 3),
 ('soccer', 3),
 ('football', 3),
 ('WWEGRR', 3),
 ('FCBarcelona', 2),
 ('YNWA', 2),
 ('cryptocurrencies', 2),
 ('Cristiano', 2),
 ('bitcoin', 2),
 ('TheBest', 2),
 ('AndresIniesta', 2),
 ('OneSeasonWonder', 2),
 ('Legend', 2),
 ('Fabregas', 2),
 ('infinit8Iniesta', 2),
 ('Mateo', 2),
 ('ICO', 2),
 ('FNH', 2),
 ('ForçaBarça', 2),
 ('Cryptocurrency', 2),
 ('bitcoins', 2),
 ('HereToCreate', 2),
 ('PSG', 2),
 ('Undertaker', 2),
 ('trademark', 2),
 ('Chelsea', 2),
 ('New_tricks', 1),
 ('La', 1),
 ('RealMadrid', 1),
 ('Dan

We can see that Iniesta shows very often in the tweet. It occured two times of the second one. So we should add this hashtag when posting.

ix.	Should	I	follow	somebody	back?

This question is just like the sixth one. We find that ForeignPolicy and fabio_adorno are somewaht alike. We also find that ForeignPolicy hasn't followed fabio_adorno. So to the scale we scrapped like around 200 per person. It's not very likely to find two with a lot of same favorites. So we think ForeignPolicy may consider follow back.

x.	What	is	the	best	time	to	post?

In [1]:
import pymysql  
import  pymysql.cursors
import numpy as np
import pandas as pd
connection = pymysql.connect(host='localhost',  
                             user='root',  
                             password='woaiyiwen16',  
                             db='sample',  
                             port=3306,  
                             charset='latin1')
cursor = connection.cursor()
sql = "select  created_at,count(*) from similarposts1 GROUP by created_at ORDER by created_at desc"
cursor.execute(sql)
result = cursor.fetchall()
result

((23, 16),
 (22, 30),
 (21, 50),
 (20, 26),
 (19, 12),
 (18, 33),
 (17, 39),
 (16, 41),
 (15, 53),
 (14, 40),
 (13, 57),
 (12, 71),
 (11, 38),
 (10, 20),
 (9, 37),
 (8, 29),
 (7, 29),
 (6, 11),
 (5, 14),
 (4, 15),
 (3, 10),
 (2, 6),
 (1, 8),
 (0, 7))

We count tweets posted per hour which using #messi and find that at noon in USA tweets raise to the highest amount. This may be caused by at that time in Spain is La Liga time. People tend to post more tweets when watching games. Also we find that many people tend to post tweets before sleep which shows as a raise at 21 o'clock.

xi.	Should	I	add	and	picture	or	url	to	my	post?

In [7]:
sql = "select count(*) from messi where text like '%http%'"
cursor.execute(sql)
result = cursor.fetchall()
result

((3,),)

We got aroung 1k tweets mentioning messi but only 3 of them attached a url.

In [11]:
from pandas import read_csv
x = read_csv('3GerardPique_facebook_statuses.csv',encoding='UTF-8')
y = x['status_type'];
y = pd.DataFrame(y)
y

status_type
0         photo
1         photo
2         photo
3         photo
4         photo
5         photo
6         photo
7         video
8          link
9          link
10        photo
11        photo
12         link
13        photo
14        photo
15        video
16        video
17        photo
18        photo
19        photo
20        video
21        video
22        video
23        video
24        photo
25        video
26        video
27        video
28        video
29        photo
..          ...
81        photo
82        video
83        photo
84        video
85        photo
86        video
87        video
88        photo
89        photo
90        photo
91        photo
92        photo
93        photo
94       status
95         link
96        photo
97        photo
98        photo
99        video
100       photo
101       photo
102       photo
103       photo
104       video
105       photo
106       photo
107       photo
108       video
109       video
110      status

[111 rows x 1 columns]

In [14]:
counts = pd.DataFrame(y['status_type'].value_counts())
counts

status_type
photo            74
video            28
status            5
link              4

In [15]:
x = read_csv('BarcelonaDNA_facebook_statuses.csv',encoding='UTF-8')
y = x['status_type'];
y = pd.DataFrame(y)
counts = pd.DataFrame(y['status_type'].value_counts())
counts

status_type
photo          1166
status          140
video            72
link              3

In [16]:
x = read_csv('LaLiga_facebook_statuses.csv',encoding='UTF-8')
y = x['status_type'];
y = pd.DataFrame(y)
counts = pd.DataFrame(y['status_type'].value_counts())
counts

status_type
video           859
photo           119
link              4
status            1

We count the use of media type on facebook. We choose the account of player, fanclub and official site. As we can see, facebook with plain text seldom appears. Most of them carry a video or a photo. So for Facebook, it's better to com up with some media tpe rather than posting a long plain text.

xii.	What’s	my	reach?

To answer this question. We decide to scrap a user's follower and search posts of each follower. By counting the tags they usually use, we try to find out that whether they are from other domains. Fot instance, we look for messi's follower and find their posts.
After a tough lookup, we got a user LuisTGS(PS4). 

In [17]:
import pandas as pd
from pandas import read_csv
import tweepy
import csv
CONSUMER_KEY = 'ilrMBG1phrX3iT3BCP6rbB0uS'
CONSUMER_SECRET = 'C2m5PwgEzgdipZ0kMlCWIvZzFtZ3Ip3y0Rbsv9PAo2wYcUBMN8'
ACCESS_TOKEN = '906874478554165250-LN01g0KGYAm2UDHgnXow3WDHxKtRFCq'
ACCESS_TOKEN_SECRET = 'Yvxbmm2OEkedq8sZ3cPHtjwPW3FVtx65JnWMtJ1drpno4'
def get_all_tweets(user_id):
    
    auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
    auth.set_access_token(ACCESS_TOKEN, ACCESS_TOKEN_SECRET)
    api = tweepy.API(auth)

    alltweets = []
    new_tweets = api.user_timeline(user_id=user_id, count=200)
    alltweets.extend(new_tweets)
    oldest = alltweets[-1].id - 1
    while len(new_tweets) > 0:
        print ("getting tweets before %s" % (oldest))
        new_tweets = api.user_timeline(user_id=user_id, count=200, max_id=oldest)
        print(new_tweets[0])
        alltweets.extend(new_tweets)
        oldest = alltweets[-1].id - 1
        print ("...%s tweets downloaded so far" % (len(alltweets)))
        outtweets = [[tweet.id_str, tweet.created_at, tweet.text.encode("utf-8"),tweet.favorite_count,tweet.retweet_count,tweet.entities.get('hashtags')] for tweet in alltweets]
        outtweets = pd.DataFrame(outtweets)
        outtweets.to_csv('%s_tweets.csv' % user_id, index=True, encoding='UTF-8')

if __name__ == '__main__':
            try:
                get_all_tweets('2518106781')
            except tweepy.TweepError as e:
                print ('Failed to run the command on that user, Skipping...')
            except IndexError as e:
                print ('List index out of range, Skipping...')

getting tweets before 988230554947674111
Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Mon Apr 23 01:35:35 +0000 2018', 'id': 988229868138782721, 'id_str': '988229868138782721', 'text': "@ItsNotNoahYT @GunsNasty 11mins is still fast reliable because you got your coins, I'm not saying it's supposed to… https://t.co/ZRsD80cB0W", 'truncated': True, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'ItsNotNoahYT', 'name': 'RT Pinned Tweet', 'id': 598849503, 'id_str': '598849503', 'indices': [0, 13]}, {'screen_name': 'GunsNasty', 'name': 'SKOL', 'id': 3967002614, 'id_str': '3967002614', 'indices': [14, 24]}], 'urls': [{'url': 'https://t.co/ZRsD80cB0W', 'expanded_url': 'https://twitter.com/i/web/status/988229868138782721', 'display_url': 'twitter.com/i/web/status/9…', 'indices': [116, 139]}]}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 988229492438192

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Thu Apr 19 22:30:45 +0000 2018', 'id': 987096188829753345, 'id_str': '987096188829753345', 'text': '@NYGfootballkev ?', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'NYGfootballkev', 'name': '0-16 tanksznpt2', 'id': 828241573, 'id_str': '828241573', 'indices': [0, 15]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 987095712776294400, 'in_reply_to_status_id_str': '987095712776294400', 'in_reply_to_user_id': 828241573, 'in_reply_to_user_id_str': '828241573', 'in_reply_to_screen_name': 'NYGfootballkev', 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'description': {'urls': []}}, 'protected': False, 'followers_co

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Sun Apr 15 20:13:35 +0000 2018', 'id': 985612118245470208, 'id_str': '985612118245470208', 'text': 'RT @NYGfootballkev: Went white water rafting today so much fucking fun', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'NYGfootballkev', 'name': '0-16 tanksznpt2', 'id': 828241573, 'id_str': '828241573', 'indices': [3, 18]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': None, 'in_reply_to_status_id_str': None, 'in_reply_to_user_id': None, 'in_reply_to_user_id_str': None, 'in_reply_to_screen_name': None, 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'description': {'urls': []}}, 'protected': False, 'followers_cou

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Fri Apr 13 02:33:41 +0000 2018', 'id': 984620611522781184, 'id_str': '984620611522781184', 'text': 'RT @Goatxyy: So Guru DMd me lol... 😂 https://t.co/KdFQTXrwkt', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'Goatxyy', 'name': '66', 'id': 975938390523830272, 'id_str': '975938390523830272', 'indices': [3, 11]}], 'urls': [], 'media': [{'id': 984590797789020160, 'id_str': '984590797789020160', 'indices': [37, 60], 'media_url': 'http://pbs.twimg.com/media/Dan4HS9VQAA5fUf.jpg', 'media_url_https': 'https://pbs.twimg.com/media/Dan4HS9VQAA5fUf.jpg', 'url': 'https://t.co/KdFQTXrwkt', 'display_url': 'pic.twitter.com/KdFQTXrwkt', 'expanded_url': 'https://twitter.com/Goatxyy/status/984590846157819909/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 640, 'h': 474, 'resize': 'fit'}, 'large': {'w': 640, 'h': 474, 'res

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Wed Apr 11 23:39:30 +0000 2018', 'id': 984214386238779392, 'id_str': '984214386238779392', 'text': 'RT @MutSoloPlug: who wants it? Im not rich like other people because i dont buy &amp; sell but i giveaway when i can https://t.co/jgExniFepk', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'MutSoloPlug', 'name': 'MutSoloPlug', 'id': 947908167161192448, 'id_str': '947908167161192448', 'indices': [3, 15]}], 'urls': [], 'media': [{'id': 984210211480088576, 'id_str': '984210211480088576', 'indices': [117, 140], 'media_url': 'http://pbs.twimg.com/media/Daid-QeVMAAXq8X.jpg', 'media_url_https': 'https://pbs.twimg.com/media/Daid-QeVMAAXq8X.jpg', 'url': 'https://t.co/jgExniFepk', 'display_url': 'pic.twitter.com/jgExniFepk', 'expanded_url': 'https://twitter.com/MutSoloPlug/status/984210214357368832/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'r

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Sun Apr 08 22:42:17 +0000 2018', 'id': 983112826221989888, 'id_str': '983112826221989888', 'text': 'RT @BryanTheMan31: i will game share MLB The Show 18 for 20-25 bucks w you. hit my dm’s. RT’s Appreciated', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'BryanTheMan31', 'name': 'Bryan (16-0)', 'id': 356548366, 'id_str': '356548366', 'indices': [3, 17]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': None, 'in_reply_to_status_id_str': None, 'in_reply_to_user_id': None, 'in_reply_to_user_id_str': None, 'in_reply_to_screen_name': None, 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'description': {'urls': []}}, 'pr

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Sat Apr 07 22:29:57 +0000 2018', 'id': 982747334437146624, 'id_str': '982747334437146624', 'text': '@_EverythingMUT @Bracey2011 @csmilinich88 @Tr3yFiv5 98 Larry Fitz', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': '_EverythingMUT', 'name': 'EverythingMUT', 'id': 740223161489313793, 'id_str': '740223161489313793', 'indices': [0, 15]}, {'screen_name': 'Bracey2011', 'name': 'Paul Bracegirdle', 'id': 235102664, 'id_str': '235102664', 'indices': [16, 27]}, {'screen_name': 'csmilinich88', 'name': 'CS', 'id': 3177752533, 'id_str': '3177752533', 'indices': [28, 41]}, {'screen_name': 'Tr3yFiv5', 'name': 'K.O.D', 'id': 622453859, 'id_str': '622453859', 'indices': [42, 51]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 982724687942701057, 'in_reply_to_status_id_str': '9827246

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Fri Apr 06 05:02:33 +0000 2018', 'id': 982121359928049665, 'id_str': '982121359928049665', 'text': '@GunsNasty He turned into a fucking lizard fuck is wrong with this dudes mentality', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'GunsNasty', 'name': 'SKOL', 'id': 3967002614, 'id_str': '3967002614', 'indices': [0, 10]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 982120683994951681, 'in_reply_to_status_id_str': '982120683994951681', 'in_reply_to_user_id': 2518106781, 'in_reply_to_user_id_str': '2518106781', 'in_reply_to_screen_name': 'LuisCal21481160', 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'descript

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Wed Apr 04 22:38:53 +0000 2018', 'id': 981662419880763392, 'id_str': '981662419880763392', 'text': 'RT @BleacherReport: in office pornography discussion taking place', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'BleacherReport', 'name': 'Bleacher Report', 'id': 890891, 'id_str': '890891', 'indices': [3, 18]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': None, 'in_reply_to_status_id_str': None, 'in_reply_to_user_id': None, 'in_reply_to_user_id_str': None, 'in_reply_to_screen_name': None, 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'description': {'urls': []}}, 'protected': False, 'followers_count': 586, '

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Tue Apr 03 04:34:08 +0000 2018', 'id': 981027045487542272, 'id_str': '981027045487542272', 'text': '@BenzoBoomin_ @CloudGodLucas @DynoHuncho Benzo just tell her u got clout and you will treat her nice 😅', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'BenzoBoomin_', 'name': 'Justin.', 'id': 973446853734948864, 'id_str': '973446853734948864', 'indices': [0, 13]}, {'screen_name': 'CloudGodLucas', 'name': 'Sober Lucas', 'id': 730126215835729921, 'id_str': '730126215835729921', 'indices': [14, 28]}, {'screen_name': 'DynoHuncho', 'name': '💀☄️dyno', 'id': 700108531324702721, 'id_str': '700108531324702721', 'indices': [29, 40]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 981026365939027968, 'in_reply_to_status_id_str': '981026365939027968', 'in_reply_to_user_id': 973446

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Sun Apr 01 15:35:17 +0000 2018', 'id': 980468651718533120, 'id_str': '980468651718533120', 'text': 'Have u opened this yet @NYGfootballkev @GunsNasty ? Is it worth https://t.co/eeN9SIkEA6', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'NYGfootballkev', 'name': '0-16 tanksznpt2', 'id': 828241573, 'id_str': '828241573', 'indices': [23, 38]}, {'screen_name': 'GunsNasty', 'name': 'SKOL', 'id': 3967002614, 'id_str': '3967002614', 'indices': [39, 49]}], 'urls': [], 'media': [{'id': 980468617182568448, 'id_str': '980468617182568448', 'indices': [64, 87], 'media_url': 'http://pbs.twimg.com/media/DZtTAziV4AA_VA9.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DZtTAziV4AA_VA9.jpg', 'url': 'https://t.co/eeN9SIkEA6', 'display_url': 'pic.twitter.com/eeN9SIkEA6', 'expanded_url': 'https://twitter.com/LuisCal21481160/status/980468651718533120/photo/1', 'type': 'ph

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Sat Mar 31 17:48:55 +0000 2018', 'id': 980139893811744774, 'id_str': '980139893811744774', 'text': '@lmaoowendy @frxaxy @hallegrace17 @SommerRay Damn', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'lmaoowendy', 'name': 'Wendy', 'id': 2240987614, 'id_str': '2240987614', 'indices': [0, 11]}, {'screen_name': 'frxaxy', 'name': 'idc', 'id': 728578937362780160, 'id_str': '728578937362780160', 'indices': [12, 19]}, {'screen_name': 'hallegrace17', 'name': 'Halle🏐🌾', 'id': 912788483084836864, 'id_str': '912788483084836864', 'indices': [20, 33]}, {'screen_name': 'SommerRay', 'name': 'Sommer Ray', 'id': 631274178, 'id_str': '631274178', 'indices': [34, 44]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 980135232589455362, 'in_reply_to_status_id_str': '980135232589455362', 'i

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Fri Mar 30 19:36:13 +0000 2018', 'id': 979804509475164161, 'id_str': '979804509475164161', 'text': '@WatkinsSzn @RynoSzn @Tr3yFiv5', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'WatkinsSzn', 'name': 'Marshall', 'id': 751192914613514240, 'id_str': '751192914613514240', 'indices': [0, 11]}, {'screen_name': 'RynoSzn', 'name': '"', 'id': 4435609513, 'id_str': '4435609513', 'indices': [12, 20]}, {'screen_name': 'Tr3yFiv5', 'name': 'K.O.D', 'id': 622453859, 'id_str': '622453859', 'indices': [21, 30]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 979800441012748289, 'in_reply_to_status_id_str': '979800441012748289', 'in_reply_to_user_id': 751192914613514240, 'in_reply_to_user_id_str': '751192914613514240', 'in_reply_to_screen_name': 'WatkinsSzn', 'user': {'id': 25181067

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Wed Mar 28 02:28:15 +0000 2018', 'id': 978821035519815685, 'id_str': '978821035519815685', 'text': '@RynoSzn Big fact', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'RynoSzn', 'name': '"', 'id': 4435609513, 'id_str': '4435609513', 'indices': [0, 8]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 978820871103176709, 'in_reply_to_status_id_str': '978820871103176709', 'in_reply_to_user_id': 4435609513, 'in_reply_to_user_id_str': '4435609513', 'in_reply_to_screen_name': 'RynoSzn', 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'description': {'urls': []}}, 'protected': False, 'followers_count': 586, 'friends_count

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Mon Mar 26 22:27:20 +0000 2018', 'id': 978398022387355648, 'id_str': '978398022387355648', 'text': '@cookieboy1794 "No pun intended" 😂', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'cookieboy1794', 'name': 'cookieboy17', 'id': 600591646, 'id_str': '600591646', 'indices': [0, 14]}], 'urls': []}, 'source': '<a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>', 'in_reply_to_status_id': 978397917491945477, 'in_reply_to_status_id_str': '978397917491945477', 'in_reply_to_user_id': 600591646, 'in_reply_to_user_id_str': '600591646', 'in_reply_to_screen_name': 'cookieboy1794', 'user': {'id': 2518106781, 'id_str': '2518106781', 'name': 'LuisTGS(PS4)', 'screen_name': 'LuisCal21481160', 'location': 'New York, USA', 'description': '#Giants\n#Playlinearmy', 'url': None, 'entities': {'description': {'urls': []}}, 'protected': False, 'f

Status(_api=<tweepy.api.API object at 0x00000221A0C8CE80>, _json={'created_at': 'Sun Mar 25 03:30:19 +0000 2018', 'id': 977749492521951233, 'id_str': '977749492521951233', 'text': 'RT @CVS_2x: RT an LIKE \nfollow @CVS_2x \n\n3 people will get 20$( new Fortnite skins ) https://t.co/6jO4dCR1pk', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'CVS_2x', 'name': '☪️♈️💲 DF', 'id': 728382593201278976, 'id_str': '728382593201278976', 'indices': [3, 10]}, {'screen_name': 'CVS_2x', 'name': '☪️♈️💲 DF', 'id': 728382593201278976, 'id_str': '728382593201278976', 'indices': [31, 38]}], 'urls': [], 'media': [{'id': 977714649306087424, 'id_str': '977714649306087424', 'indices': [85, 108], 'media_url': 'http://pbs.twimg.com/media/DZGKSxuVwAAuPoH.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DZGKSxuVwAAuPoH.jpg', 'url': 'https://t.co/6jO4dCR1pk', 'display_url': 'pic.twitter.com/6jO4dCR1pk', 'expanded_url': 'https://twitter.com/CVS_2x/status/97771

List index out of range, Skipping...


In [21]:
import csv
import pandas as pd
import sys, getopt, pprint
from pymongo import MongoClient
user_id = '2518106781'
csvfile = open('%s_tweets.csv' % user_id,'r')


In [2]:
sql = "SELECT tag1, count(*) FROM gai2518106781_tweets group by tag1 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('ThunderUp', 6),
 ('Fortnite', 4),
 ('UltimateLegends', 4),
 ('ComingTomorrow', 3),
 ('PS4share', 3),
 ('NFLDraft', 2),
 ('Giants', 2),
 ('MUTSeason', 2),
 ('BossBattles', 2),
 ('SignatureSeries', 2),
 ('GiantsPride', 1),
 ('SaquonBarkley', 1),
 ('JeopardySports', 1),
 ('saquonszn', 1),
 ('StayTuned', 1),
 ('PlayLineArmy', 1),
 ('REPOST', 1),
 ('NYGiants', 1),
 ('MegaMonday', 1),
 ('RockDeals', 1),
 ('MaddenBowl', 1),
 ('IntelRigChallenge', 1),
 ('MUTCommunity', 1),
 ('UltimateLegend', 1),
 ('tictocnews', 1),
 ('Columbine', 1),
 ('WorldCup', 1),
 ('GTAOnline', 1),
 ('Deadpool2', 1),
 ('WhyNot', 1),
 ('ForPaul', 1),
 ('PlayoffMoments', 1),
 ('ShirtlessSwag', 1),
 ('KeepPounding', 1),
 ('Thunder', 1),
 ('MUT18', 1),
 ('STAR', 1),
 ('NFLReplays', 1))

In [3]:
sql = "SELECT tag2, count(*) FROM gai2518106781_tweets group by tag2 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 36),
 ('MUTSeason', 12),
 ('NFLDraft', 3),
 ('SaquonBarkley', 1),
 ("'Legit", 1),
 ('sweepstakes', 1),
 ('Falcons', 1),
 ('ComingTomorrow', 1),
 ('NBAPlayoffs', 1),
 ('MADDEN18', 1))

In [4]:
sql = "SELECT tag3, count(*) FROM gai2518106781_tweets group by tag3 order by count(*) desc"
cursor.execute(sql)
result = cursor.fetchall()
result

(('', 56), ('SaquonBarkley', 1), ('giveaways', 1))

We can see this follower is a fan of gaming and we can assume he is one of the game domain. A follower can see a user's post immediately. So this follower here means the user's reach to the game domain. To get an entire reach, we have to look up all the followers and those who retweet or like but did not follow.

License (MIT)

Copyright (c) 2018 

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.